### ISSUE TITLE
Saving as PXT file

### ISSUE THREAD

**Issue by damianburrin** [Microsoft/pxt#769]
Have I missed something.  Was there a reason for the save as PXT option being removed when PXT.IO was merged into "codethemicrobit" and the microbit.org? 

This for me was a really useful feature as it meant the file could be saved locally but not as a HEX file

**Comment by abchatra** [comment]
@damianburrin Hex file does contain the entire project information as well. Can you please elaborate on the scenario you are looking to address? Are you looking for a way to save the readable typescript file?

**Comment by damianburrin** [comment]
Two points really.
One i used to be able to do it on pxt.io as i have some pxt files here so it was a feature that i used to use and wondered why i can't now.
Second in the (our) school envirnoment hex and bin files are blocked.  To be fair so are USB drives.  We've had usb ports unlocked for microbits only which is easy as they all have the same uid.  However hex files for us a restricted to drive G (the microbit drive) so students can't save work for later review.
As i mention above.  I have saved PXT files in the past so this feature must have been somewhere as an export/import feature.
Damian
On 14 Nov 2016 23:11, Abhijith Chatra <notifications@github.com> wrote:@damianburrin Hex file does contain the entire project information as well. Can you please elaborate on the scenario you are looking to address? Are you looking for a way to save the readable typescript file?

**Comment by abchatra** [comment]
Save project is back and live. You can go Projects->SaveProjects

**Comment by damianburrin** [comment]
Unless i'm missing something this doesn't address the original query/problem.  The save feature is there but saves as HEX.  The PXT.IO used to have a save as PXT option.  I could do with this option being put back.  As i mentioned previously we are not allowed to save HEX/BIN to the network drive other than drive G which is the Micro:bit so students can't keep a copy of part completed work or finished projects anymore.

**Comment by pelikhan** [comment]
So .pxt file can be saved but .hex are blocked? Is my understanding correct? What about zip files?

**Comment by OwenBrotherwood** [comment]
Note: In many situations, a zip file is opened by, for example, anti virus. Any banned extension is still banned, even if zipped.

**Comment by OwenBrotherwood** [comment]
At this time, my recommendation is "ping pong" on that a archive content of pxt.json and the files referenced by pxt.json is the broadest fit as to what is needed

**Comment by pelikhan** [comment]
We are considering zipping the sources of the project which would contain all the necessary files, .blocks, .ts, .json. We are not planning to zip the .hex file.

**Comment by damianburrin** [comment]
Corporate policy blocks by extension and location.  You don't want to know the hoops i had to jump through to allow usb devices at all and then hex/bin files to a specific drive.  PXT doesn't appear on their radar so that was ideal.  zip is allowed for staff not students.On 10 Dec 2016 17:46, Peli de Halleux <notifications@github.com> wrote:So .pxt file can be saved but .hex are blocked? Is my understanding correct? What about zip files?

**Comment by damianburrin** [comment]
Sorry i don't followOn 10 Dec 2016 17:52, Owen Brotherwood <notifications@github.com> wrote:At this time, my recommendation is "ping pong" on that a archive content of pxt.json and the files referenced by pxt.json is the broadest fit as to what is needed

**Comment by damianburrin** [comment]
Why was the pxt extension removed?On 10 Dec 2016 18:03, Peli de Halleux <notifications@github.com> wrote:We are considering zipping the sources of the project which would contain all the necessary files, .blocks, .ts, .json. We are not planning to zip the .hex file.

**Comment by OwenBrotherwood** [comment]
Please note: 
I stated archive format, a way of placing more than one file into one file.
I have several roles/jobs in my life, enforcing policies concerning files and circumventing them dependant on.
@damianburrin has now correctly stated what was needed in the ping/pong concerning archive format as my next direct question to him was if his installation could accept .zip with fies of the stated extenstion.
I was worried he could not.

**Comment by OwenBrotherwood** [comment]
.pxt seems to be a bad a choice as any. 
Compression is not an issue as such: and I am an old unix guy that used to exchange binary in ascii format. This is an easy case, exchange text files in text file.
Concat files to one file and extension .pxt is a fit for damian.
Allows for quick exchange of programs via any copy and paste service on the internet.

**Comment by damianburrin** [comment]
I'm still rather confused.  PXT was a previous feature.  We used it successfully.  Was there a technical reason for removal in the first place?

**Comment by pelikhan** [comment]
We removed it to avoid having to support 2 file formats. Obviously we were not aware of all the implications.

**Comment by damianburrin** [comment]
ah - i see.

In my experience in schools.  IT support (especially if like me your part of a large academy group) get very twitchy of binary file formats.  Ours even blocked mp3/4 and Jpeg picutres files as they felt they were dangerous to the security of the network.  It took quite a but of well what do you expect the music/media/art departments to do!! The standard answer is always no! and we are always looking for ways to support our kids against these frustrations.

**Comment by OwenBrotherwood** [comment]
What other purpose doe the hex format help?
Supporting two formats is no joke.
Is it possible that the hex format is not really necessary.

**Comment by damianburrin** [comment]
Isn't it the hex file that is flashed to the microbit?  I guess thats required.  Its a shame to lose local file saving but i guess thats the way it will go for us.  i appreciate that you can't make major changes to the programs  route plan.  On 11 Dec 2016 11:39, Owen Brotherwood <notifications@github.com> wrote:What other purpose doe the hex format help?
Supporting two formats is no joke.
Is it possible that the hex format is not really necessary.

**Comment by OwenBrotherwood** [comment]
Probably :)
So given the use of the hex file is to be able to save a result that can also be used to program the micro:bit without access to pxt, then the use case for the hex is "no access to pxt" and need to program a micro:bit with a previously saved hex can be present.
From the hex format, it is apparently possible to load the hex file to recreate the pxt block/javascript, the reason for only using hex (support of 2 file formats) was a good enough reason except for the use case presented by yourself.
My own use case is a file that contains all the files needed in plain text (with no need for compression).
Luckily, there is no XML in the files needed as far as I have found :)

**Comment by damianburrin** [comment]
I wonder what other schools do. Though TBH through my work with CAS many schools shelved microbits as they came out far to late so they were just passed on.  its taking a lot of work to try and get local schools near me to buy in and use them.  I know most of these are using MU and python as it supports the ks3 and 4 curriculum and its the path of least resistance for security policiesOn 11 Dec 2016 13:39, Owen Brotherwood <notifications@github.com> wrote:Probably :)
So given the use of the hex file is to be able to save a result that can also be used to program the micro:bit without access to pxt, then the use case for the hex is "no access to pxt" and need to program a micro:bit with a previously saved hex can be present.
From the hex format, it is apparently possible to load the hex file to recreate the pxt block/javascript, the reason for only using hex (support of 2 file formats) was a good enough reason except for the use case presented by yourself.
My own use case is a file that contains all the files needed in plain text (with no need for compression).
Luckily, there is no XML in the files needed as far as I have found :)

**Comment by pelikhan** [comment]
Could you elaborate on the path of least resistance?

**Comment by damianburrin** [comment]
@pelikhan - Sorry I'm not suggesting the PXT community are resisting, what i mean is getting an agreement allowing any coding was difficult (C was bared out right Python could be demonstrated as the exam board prefrred) be we could demonstrate a requirement.  Getting permission to have USB devices was a straight no but we were able to demonstrate a requirement for the Micro:bits so this was lifted slightly to enable this device.  This was only done as the the Micro:bits have the same GUID so this could be centrally allowed.

I have managed to get HEX/BIN files allowed as long as they are directly to the G drive (this is where the microbit maps to) There is no movement on this.  I am part of a very large academy chain so this is the same for all schools in our chain.

The "path of least resistance" is what we can get allowed by core IT that doesn't require me to bang my head against the wall.

Currently this would be - using PXT (which for some reason needs allowing through our filtering system! and the emulator doesn't get allowed through) for basic introdcuctions to the Micro:bit and coding where students can just upload code and be complex enough that would require continuation to the next lesson and anything more complex taught through MicroPython as the saving as Py files is already in place.

**Comment by pelikhan** [comment]
Is this assessment correct: since you have to through IT to download files **anyway**, you might as well request to install a native editor (eg: Mu) and get auto-flash, file saving in one shot. Is this correct?

The simulator sub-domain is blocked? That's quite unfortunate. Any reasons behind it?

**Comment by abchatra** [comment]
@damianburrin are .zip files allowed?

**Comment by damianburrin** [comment]
Mu is Python.  As python is allowed it causes no issues with the file type blocks so everything works.  when we use the online python  editor its just text so copy and paste to notepad and save as a python file.  It works.We used to save as PXT, this for some reason didn't raise any flags so was allowed.  We requested it as a program specific file type.  I tried that with hex and was shut down as hex is also used for flashing hardware devices im told.As for the emulator.  No one seems to know why its being filtered out unless its using a port thats blocked on our firewall.  All addresses and sub domains for codethemicrobit and pxt.io are set to allow.  it just doesn't on our corperate network.On 11 Dec 2016 18:02, Peli de Halleux <notifications@github.com> wrote:Is this assessment correct: since you have to through IT to download files anyway, you might as well request to install a native editor (eg: Mu) and get auto-flash, file saving in one shot. Is this correct?
The simulator sub-domain is blocked? That's quite unfortunate. Any reasons behind it?

**Comment by damianburrin** [comment]
Not for students Abhijith.As someone mentioned earlier a I think the software is quite intelligent. Ive tried to zip files in the past and it failed so it must open them and checkOn 11 Dec 2016 18:03, Abhijith Chatra <notifications@github.com> wrote:@damianburrin are .zip files allowed?

**Comment by pelikhan** [comment]
@damianburrin this is very insightfull. Unfortunately, we do not get to hear these experiences through the level of indirection between us and the issues faced in the classroom.

**Comment by tballmsft** [comment]
@pelikhan - any more insight on this topic?

**Comment by pelikhan** [comment]
@abchatra

**Comment by abchatra** [comment]
There are two workaround for this issue:
1. Share feature (this allows the students to share their project with anyone)
2. Manually copying the typescript code to a text file.

This point of time we don't intend to change the extension of the file as we don't hear similar feedback from others. Do let us know if bot of these workarounds don't work in school settings.

**Comment by damianburrin** [comment]
It's a shame that a feature is removed but I appreciate your limited in what can
be done.  I think though you are over estimating what happens in UK schools when
you say "change the extension of the file as we don't hear similar feedback from
others"

The reason you don't hear from other schools is mostly teachers are not
technical if a feature isn't there they just won't use it.  The fact the feature
was there and then removed by MS prompted my question.  If it hadn't been there
before I'd not have thought to even ask.

By not having this feature you limit the use of the software in all UK schools.
 You will find PXT is just not being used in schools.  Schools do not like the
"share feature" it causes us to worry about students communicating with
inappropriate adults so we just block it.  School networks block binary files so
saving ofline as HEX just doesn't happen.  The schools that do make use of
microbits applied a work around to only allow the saving of these files to the
micro:bit drive itself.

As I say I appreciate you don't feel the need to re-apply this feature but I
think you need to look more closely as to what is happening in schools and the
restrictions we work under.  Just because your not getting lots of people asking
doesn't mean they are happy and using it, more likely they are just not using
it.


Thanks

Damian

**Comment by pelikhan** [comment]
@jaustin what's the Foundation POV here?

**Comment by abchatra** [comment]
I have reopened the issue. 

We will not go back to .pxt extension though:

- .pxt is still a binary format and still be blocked in the schools
- PXT is not a consumer facing name of the project (Microsoft MakeCode is).
- There is an issue of explaining to the users multiple formats .pxt & .hex which essentially does the same thing today.

I do acknowledge the fact that, it is possible this scenario is important for some schools. We will work on an holistic approach for those. There are other formats which might solve the issue.

**Comment by abchatra** [comment]
Closing old issues. If this is still an issue please reopen.
